Add OpenPMIx for OFI and UCX machine layers (#3194)
UIUC-PPL/charm: doc/charm++/manual.rst

^^^
UCX stands for Unified Communication X and is a high performance communication
library that can be used as a backend networking layer for Charm++ builds on
supported transports like Mellanox Infiniband, Intel Omni-Path, Cray GNI, TCP/IP, etc.

> In order to install Charm++ with the UCX backend, you require UCX or HPC-X modules

Reviewer: ```suggestion
In order to install Charm++ with the UCX backend, UCX or HPC-X modules are required
```

> in your environment. In case UCX or HPC-X is not available in your environment,

Reviewer: ```suggestion
in your environment. In case UCX and HPC-X are not available in your environment,
```

After installing UCX, there are several supported process management interfaces (PMI)
that can be specified as options in order to build Charm++ with UCX. These include

> Simple PMI, Slurm PMI, Slurm PMI 2 and PMIx (included in OpenMPI or OpenPMIx). Currently, in order to

Reviewer: ```suggestion
Simple PMI, Slurm PMI, Slurm PMI 2, and PMIx (included in OpenMPI or OpenPMIx). Currently, in order to
```

.. code-block:: bash

   $ wget https://github.com/libevent/libevent/releases/download/release-2.1.12-stable/libevent-2.1.12-stable.tar.gz
   $ tar -xf libevent-2.1.12-stable.tar.gz
   $ cd libevent-2.1.12-stable
   $ ./configure --prefix=$HOME/libevent-2.1.12-stable/build
   $ make -j
   $ make install

Then you can download and build OpenPMIx as follows:

.. code-block:: bash

   $ wget https://github.com/openpmix/openpmix/releases/download/v3.1.5/pmix-3.1.5.tar.gz
   $ tar -xf pmix-3.1.5.tar.gz
   $ cd pmix-3.1.5
   $ ./configure --prefix=$HOME/pmix-3.1.5/build --with-libevent=$HOME/libevent-2.1.12-stable/build
   $ make -j
   $ make install

Finally, Charm++ with the UCX backend can be built with OpenPMIx using the following command
(with the OpenPMIx installation passed as an additional ``--basedir`` argument):

.. code-block:: bash

   $ ./build charm++ ucx-linux-x86_64 openpmix --with-production --enable-error-checking --basedir=$HOME/ucx/build --basedir=$HOME/pmix-3.1.5/build -j

It should be noted that in the absence of a working launcher such as ``jsrun``, an MPI distribution
such as OpenMPI may also be required to run Charm++ applications built with the UCX backend

> and OpenPMIx. Since you no longer need PMIx included with OpenMPI, any verison of OpenMPI can be built

Reviewer: ```suggestion
and OpenPMIx. Since you no longer need PMIx included with OpenMPI, any version of OpenMPI can be built
```

(including v4.0.5 and later) with your build of OpenPMIx using the ``--with-pmix`` flag,
such as the following:

.. code-block:: bash

   $ wget https://download.open-mpi.org/release/open-mpi/v4.0/openmpi-4.0.5.tar.gz
   $ tar -xf openmpi-4.0.5.tar.gz
   $ cd openmpi-4.0.5
   $ ./configure --enable-mca-no-build=btl-uct --with-pmix=$HOME/pmix-3.1.5/build --prefix=$HOME/openmpi-4.0.5/build
   $ make -j
   $ make install

Before executing a Charm++ program, you may need to check that ``LD_LIBRARY_PATH`` and ``PATH``
are set to include OpenPMIx (and OpenMPI, if needed).

> To use PMIx included with OpenMPI, you would require an OpenMPI implementation with PMIx
> enabled to be installed on your system. As a reminder, PMIx is no longer included in OpenMPI

Reviewer: ```suggestion
To use PMIx included with OpenMPI, an OpenMPI implementation with PMIx enabled is required
to be installed on your system. As a reminder, PMIx is no longer included in OpenMPI
```
